File: Example/ObjC/Bulletin/TextFieldBulletinPage.m
/**
 *  BulletinBoard
 *  Copyright (c) 2017 - present Alexis Aubry. Licensed under the MIT license.
 */

#import "TextFieldBulletinPage.h"

@interface TextFieldBulletinPage () <UITextFieldDelegate>

@property (nonatomic, strong, readwrite) UITextField *textField;

@end

@implementation TextFieldBulletinPage

- (NSArray<UIView *> *)makeViewsUnderDescriptionWithInterfaceBuilder:(BLTNInterfaceBuilder *)interfaceBuilder
{
    self.textField = [interfaceBuilder makeTextFieldWithPlaceholder:@"Full Name"
                                                          returnKey:UIReturnKeyJoin
                                                           delegate:self];
    return @[self.textField];
}

- (void)onDisplay
{
    // Uncomment to start typing when the bulletin item is presented
    // [self.textField becomeFirstResponder];
}

- (void)tearDown
{
    [self.textField setDelegate:nil];
}

#pragma mark UITextFieldDelegate

- (void)actionButtonTappedWithSender:(UIButton *)sender
{
    [self textFieldShouldReturn:self.textField];
    [super actionButtonTappedWithSender:sender];
}

- (BOOL)isValidInput:(NSString *)text
{
    return text != NULL && text.length > 0;
}

- (BOOL)textFieldShouldReturn:(UITextField *)textField
{
    [self.textField resignFirstResponder];
    return YES;
}

- (BOOL)textFieldShouldEndEditing:(UITextField *)textField
{
    return YES;
}

- (void)textFieldDidEndEditing:(UITextField *)textField
{
    if ([self isValidInput:textField.text]) {
        if (self.textInputHandler) {
            self.textInputHandler(self, textField.text);
        }
    } else {
        UIColor *redColor = [[UIColor redColor] colorWithAlphaComponent:0.3];
        self.descriptionLabel.textColor = [UIColor redColor];
        self.descriptionLabel.text = @"You must enter a valid name to continue.";
        self.textField.backgroundColor = redColor;
    }
}

@end

File: Example/Swift/Bulletin/TextFieldBulletinPage.swift
/**
 *  BulletinBoard
 *  Copyright (c) 2017 Alexis Aubry. Licensed under the MIT license.
 */

import UIKit
import BLTNBoard

/**
 * An item that displays a text field.
 *
 * This item demonstrates how to create a bulletin item with a text field and how it will behave
 * when the keyboard is visible.
 */

class TextFieldBulletinPage: FeedbackPageBLTNItem {

    @objc public var textField: UITextField!

    @objc public var textInputHandler: ((BLTNActionItem, String?) -> Void)? = nil

    override func makeViewsUnderDescription(with interfaceBuilder: BLTNInterfaceBuilder) -> [UIView]? {
        textField = interfaceBuilder.makeTextField(placeholder: "First and Last Name", returnKey: .done, delegate: self)
        return [textField]
    }

    override func tearDown() {
        super.tearDown()
        textField?.delegate = nil
    }

    override func actionButtonTapped(sender: UIButton) {

        if textFieldShouldReturn(self.textField) {
            textInputHandler?(self, textField.text)
            super.actionButtonTapped(sender: sender)
        }

    }

}

// MARK: - UITextFieldDelegate

extension TextFieldBulletinPage: UITextFieldDelegate {

    @objc open func isInputValid(text: String?) -> Bool {

        if text == nil || text!.isEmpty {
            return false
        }

        return true

    }

    func textFieldShouldReturn(_ textField: UITextField) -> Bool {

        if isInputValid(text: textField.text) {

            textField.resignFirstResponder()
            textInputHandler?(self, textField.text)
            return true

        } else {

            descriptionLabel!.textColor = .red
            descriptionLabel!.text = "You must enter some text to continue."
            textField.backgroundColor = UIColor.red.withAlphaComponent(0.3)
            return false

        }

    }

}
